lib/autoupdate/agent/installer.go

```diff
@@ -0,0 +1,317 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package agent
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"text/template"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	checksumType   = "sha256"
+	checksumHexLen = sha256.Size * 2 // bytes to hex
+)
+
+// LocalInstaller manages the creation and removal of installations
+// of Teleport.
+type LocalInstaller struct {
+	// InstallDir contains each installation, named by version.
+	InstallDir string
+	// HTTP is an HTTP client for downloading Teleport.
+	HTTP *http.Client
+	// Log contains a logger.
+	Log *slog.Logger
+	// ReservedFreeTmpDisk is the amount of disk that must remain free in /tmp
+	ReservedFreeTmpDisk uint64
+	// ReservedFreeInstallDisk is the amount of disk that must remain free in the install directory.
+	ReservedFreeInstallDisk uint64
+}
+
+// Remove a Teleport version directory from InstallDir.
+// This function is idempotent.
+func (li *LocalInstaller) Remove(ctx context.Context, version string) error {
+	versionDir := filepath.Join(li.InstallDir, version)
+	sumPath := filepath.Join(versionDir, checksumType)
+
+	// invalidate checksum first, to protect against partially-removed
+	// directory with valid checksum.
+	err := os.Remove(sumPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return trace.Wrap(err)
+	}
+	if err := os.RemoveAll(versionDir); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Install a Teleport version directory in InstallDir.
+// This function is idempotent.
+func (li *LocalInstaller) Install(ctx context.Context, version, template string, flags InstallFlags) error {
+	versionDir := filepath.Join(li.InstallDir, version)
+	sumPath := filepath.Join(versionDir, checksumType)
+
+	// generate download URI from template
+	uri, err := makeURL(template, version, flags)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get new and old checksums. If they match, skip download.
+	// Otherwise, clear the old version directory and re-download.
+	checksumURI := uri + "." + checksumType
+	newSum, err := li.getChecksum(ctx, checksumURI)
+	if err != nil {
+		return trace.Errorf("failed to download checksum from %s: %w", checksumURI, err)
+	}
+	oldSum, err := readChecksum(sumPath)
+	if err == nil {
+		if bytes.Equal(oldSum, newSum) {
+			li.Log.InfoContext(ctx, "Version already present.", "version", version)
+			return nil
+		}
+		li.Log.WarnContext(ctx, "Removing version that does not match checksum.", "version", version)
+		if err := li.Remove(ctx, version); err != nil {
```

Comment:
Note that this will remove the active installation if the checksum does not match. In the next PR, `Remove` will error if we try to remove a linked (and therefore active) installation.